curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md

```diff
@@ -30,14 +30,76 @@ tests:
+        assert.equal(data1.returnNum, 3.78541);
+        assert.equal(data1.returnUnit, 'l');
+        const data2 = await $.get(getUserInput('url') + '/api/convert?input=10gal');
+        assert.equal(data2.returnNum, 37.8541);
+        assert.equal(data2.returnUnit, 'l');
+        const data3 = await $.get(getUserInput('url') + '/api/convert?input=1l');
+        assert.equal(data3.returnNum, 0.26417);
+        assert.equal(data3.returnUnit, 'gal');
+        const data4 = await $.get(getUserInput('url') + '/api/convert?input=10l');
+        assert.equal(data4.returnNum, 2.64172);
+        assert.equal(data4.returnUnit, 'gal');
+      } catch(xhr) {
+        throw new Error(xhr.responseText);
+      }
+    }
+    "
+  - text: I can convert <code>'lbs'</code> to <code>'kg'</code> and vice versa. (1 lbs to 0.453592 kg)
+    testString: "async getUserInput => { 
+      try {
+        const data1 = await $.get(getUserInput('url') + '/api/convert?input=1lbs');
+        assert.equal(data1.returnNum, 0.45359);
+        assert.equal(data1.returnUnit, 'kg');
+        const data2 = await $.get(getUserInput('url') + '/api/convert?input=10lbs');
+        assert.equal(data2.returnNum, 4.53592);
+        assert.equal(data2.returnUnit, 'kg');
+        const data3 = await $.get(getUserInput('url') + '/api/convert?input=1kg');
+        assert.equal(data3.returnNum, 2.20462);
+        assert.equal(data3.returnUnit, 'lbs');
+        const data4 = await $.get(getUserInput('url') + '/api/convert?input=10kg');
+        assert.equal(data4.returnNum, 22.04624);
+        assert.equal(data4.returnUnit, 'lbs');
+      } catch(xhr) {
+        throw new Error(xhr.responseText);
+      }
+    }
+    "
+  - text: I can convert <code>'mi'</code> to <code>'km'</code> and vice versa. (1 mi to 1.60934 km)
+    testString: "async getUserInput => { 
+      try {
+        const data1 = await $.get(getUserInput('url') + '/api/convert?input=1mi');
+        assert.equal(data1.returnNum, 1.60934); 
+        assert.equal(data1.returnUnit, 'km');
+        const data2 = await $.get(getUserInput('url') + '/api/convert?input=10mi');
+        assert.equal(data2.returnNum, 16.0934); 
+        assert.equal(data2.returnUnit, 'km');
+        const data3 = await $.get(getUserInput('url') + '/api/convert?input=1km');
+        assert.equal(data3.returnNum, 0.62137); 
+        assert.equal(data3.returnUnit, 'mi');
+        const data4 = await $.get(getUserInput('url') + '/api/convert?input=10km');
+        assert.equal(data4.returnNum, 6.21373); 
+        assert.equal(data4.returnUnit, 'mi');
+      } catch(xhr) {
+        throw new Error(xhr.responseText);
+      }
+    }
+    "
+  - text: If my unit of measurement is invalid, returned will be <code>'invalid unit'</code>.
+    testString: "async getUserInput => { 
+      try {
+        const data = await $.get(getUserInput('url') + '/api/convert?input=1min');
+        assert.equal(data.initUnit, 'invalid unit'); 
```

Comment:
```suggestion
        assert(data.initUnit === 'invalid unit' || data === 'invalid unit');
```

Here's the only problem I see. Not sure what the actual intended return was when this project was created, but it doesn't specifically say to put that text in the json object. So I think this should account for either way.

Reply:
I was not too sure either, but the inner-app functional tests fail, if an object is **not** returned with the respective fields.

Reply:
[The boilerplate](https://github.com/freeCodeCamp/boilerplate-project-metricimpconverter/blob/gomix/tests/2_functional-tests.js) shows the tests empty, so the user can write them either way I think. I feel like the instructions aren't clear enough to not allow this to pass, so I vote to add it.

> the inner-app functional tests fail, if an object is not returned with the respective fields.

I'm not sure I see it, where?

Reply:
This is in the boilerplate:
```
test('Convert 10L (valid input)', function(done) {
       chai.request(server)
        .get('/api/convert')
        .query({input: '10L'})
        .end(function(err, res){
          assert.equal(res.status, 200);
          assert.equal(res.body.initNum, 10);
          assert.equal(res.body.initUnit, 'L');
          assert.approximately(res.body.returnNum, 2.64172, 0.1);
          assert.equal(res.body.returnUnit, 'gal');
          done();
        });
      });
```
So, if those properties of `res.body` are not present, errors will be thrown. This is in `tests/2_functional-tests.js`

I agree, it would be best to be explicit about the return format required, but that is how I inferred how to complete my project - through the example test.

Reply:
Thanks @Sky020 😄